From adebfcf863676e658df57e140c06d482ddd52a4b Mon Sep 17 00:00:00 2001
From: Guy Nisan <gnisan@habana.ai>
Date: Wed, 27 Jan 2021 15:57:57 +0200
Subject: [PATCH] [SW-30454] i2c: calculate pec for read data

I2C (SMBus) defines a way to calculate PEC. The algorithm is CRC-8-CCITT
which is defined by polynomial (0x07) x^8 + x^2 + x + 1.

The PEC is calculated over all the data and addresses from start to stop
including read and write bit.
e.g. for slave address 0x4f -> (0x4f << 1) | is_read -> 0x9e / 0x9f

          w        w       w       r          r      r          r
Master -[0x9e]->-[CMD]-|
        ADDR-W         |-[0x9f]->-[D]-...->-[PEC]->-[D]-...->-[PEC]
                         ADDR-R	    data              data

When master receives a PEC different than the one it calculated, it is
expected it will abort the operation and ignore the data.

Change-Id: I85c7022c2442cc1e798f021e66b5d18b2b3449fd
Signed-off-by: Guy Nisan <gnisan@habana.ai>
---

diff --git a/common/cmd_module.h b/common/cmd_module.h
index bce61db..ff63686 100644
--- a/common/cmd_module.h
+++ b/common/cmd_module.h
@@ -100,6 +100,7 @@
 void module_i2c_s_setup(void);
 enum cmd_check_res module_i2c_s_check(void);
 void module_i2c_s_disable(void);
+void module_i2c_s_enable(void);
 void module_i2c_s_run(void);
 void module_i2c_s_run_fast(void);
 void module_i2c_s_teardown(void);
diff --git a/common/module_i2c_slave.c b/common/module_i2c_slave.c
index 9c6ee24..1d7fea4 100644
--- a/common/module_i2c_slave.c
+++ b/common/module_i2c_slave.c
@@ -14,6 +14,7 @@
 #include <preboot.h>
 #include <preboot/thermal.h>
 #include <hl_i2c_slave_events.h>
+#include <crc8.h>
 
 #include "cmd_module.h"
 #include "pll.h"
@@ -31,6 +32,7 @@
 	u8 cmd;
 	u8 seq_cmd;
 	u8 cnt;
+	bool is_sync_error;
 };
 
 struct i2cslave_data {
@@ -38,15 +40,15 @@
 	struct i2cslave_buf buf;
 };
 
-static struct i2cslave_data i2cs_data = {
-	.buf = { .data = &i2cs_data.buf.cmd },
-	.seq = { .wr = 1 },
-};
+static struct i2cslave_data i2cs_data;
 
 static void i2cslave_reset_buffer(struct i2cslave_buf *buf)
 {
 	buf->data = &i2cs_data.buf.cmd;
 	buf->cnt = 0;
+	buf->is_sync_error = false;
+	i2cslave_pec = 0;
+	i2cslave_pec_digest(I2C_SLAVE_ADDRESS << 1);
 }
 
 static void i2cslave_init(bool debug_mode)
@@ -142,7 +144,7 @@
 		module_i2c_s_run();
 }
 
-void module_i2c_s_setup(void)
+void module_i2c_s_enable(void)
 {
 	bool debug_mode;
 
@@ -157,6 +159,12 @@
 	}
 
 	preboot_puts_dbg("I2C SLAVE initialized\n");
+}
+
+void module_i2c_s_setup(void)
+{
+	i2cslave_pec_init();
+	module_i2c_s_enable();
 }
 
 enum cmd_check_res module_i2c_s_check(void)
@@ -220,16 +228,17 @@
 
 static inline u8 *i2cslave_consume_hw(u8 *data)
 {
+	register u8 val = (u8)regmap_readl(mmPSOC_I2C_S_IC_DATA_CMD);
+	bool is_virt;
+
 	if (data) {
-		*data = (u8)regmap_readl(mmPSOC_I2C_S_IC_DATA_CMD);
+		*data = (u8)val;
 		data++;
-	} else {
-		/*
-		 * No data buffer - just consume. This may happens when the
-		 * upper layer expect a request data but does not need it.
-		 */
-		(void)regmap_readl(mmPSOC_I2C_S_IC_DATA_CMD);
 	}
+
+	is_virt = I2CSLAVE_IS_VIRTUAL_CMD(i2cs_data.buf.cmd);
+	if (!is_virt)
+		i2cslave_pec_digest(val);
 
 	return data;
 }
@@ -244,13 +253,8 @@
 	/* Notify upper layer */
 	seq_wr = i2cslave_handle_write_req(buf->cmd, &buf->data, buf->cnt);
 
-	if (seq_wr) {
+	if (seq_wr)
 		buf->cnt++;
-	} else {
-		/* Upper layer finished it's process, reset buffer */
-		i2cslave_reset_buffer(buf);
-		seq_wr = 1;
-	}
 
 	return seq_wr;
 }
@@ -264,6 +268,12 @@
 	seq.rd = 0;
 
 	do {
+		/* Upper layer finished it's process, reset buffer */
+		if (!seq.wr) {
+			i2cslave_reset_buffer(&i2cs_data.buf);
+			seq.wr = 1;
+			data = i2cs_data.buf.data;
+		}
 		/* Consume data as long as we expect it */
 		if (seq.wr) {
 			seq.wr--;
@@ -285,22 +295,31 @@
 {
 	struct i2cslave_buf *buf = &i2cs_data.buf;
 	u8 seq_cmd = buf->seq_cmd;
-	u8 ret = i2cslave_handle_read_req(seq_cmd, i2cs_data.seq.rd);
+	u8 ret;
+
+	/* Check if we are in the middle of write request processing */
+	if (i2cs_data.seq.wr) {
+		preboot_puts("I2C SLAVE: sync error\n");
+		i2cslave_handle_sync_error();
+		i2cs_data.seq.wr = 0;
+		buf->is_sync_error = true;
+	}
+
+	if (buf->is_sync_error)
+		return 0;
+
+	ret = i2cslave_handle_read_req(seq_cmd, i2cs_data.seq.rd);
+
+	if (!i2cs_data.seq.rd)
+		i2cslave_pec_digest((I2C_SLAVE_ADDRESS << 1) | 1);
 
 	/* For next sequential read */
 	i2cs_data.seq.rd++;
-	if (!I2CSLAVE_IS_VIRTUAL_CMD(seq_cmd))
+	if (!I2CSLAVE_IS_VIRTUAL_CMD(seq_cmd)) {
 		i2cs_data.buf.seq_cmd = seq_cmd + 1;
-
-	/* Check if we are in the middle of write request processing */
-	if (i2cs_data.buf.cnt) {
-		preboot_puts("I2C SLAVE: sync error\n");
-		i2cslave_handle_sync_error();
-
-		/* Reset buffer */
-		i2cs_data.seq.wr = 1;
-		i2cslave_reset_buffer(buf);
+		i2cslave_pec_digest(ret);
 	}
+
 	return ret;
 }
 
diff --git a/common/module_i2c_slave.h b/common/module_i2c_slave.h
index 6a3af95..a5ebb84 100644
--- a/common/module_i2c_slave.h
+++ b/common/module_i2c_slave.h
@@ -28,4 +28,9 @@
 bool i2cslave_is_reset_dev(void);
 void i2cslave_reset_dev(void);
 
+extern u8 i2cslave_pec;
+
+void i2cslave_pec_init(void);
+void i2cslave_pec_digest(u8 val);
+
 #endif /* _MODULE_I2C_SLAVE_H_ */
diff --git a/include/crc8.h b/include/crc8.h
new file mode 100644
index 0000000..e6d9446
--- /dev/null
+++ b/include/crc8.h
@@ -0,0 +1,16 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2017-2020 HabanaLabs Ltd.
+ * All Rights Reserved.
+ *
+ */
+
+#ifndef _CRC8_H_
+#define _CRC8_H_
+
+#define CRC8_TABLE_SIZE		256
+
+void crc8_populate_msb(u8 table[CRC8_TABLE_SIZE], u8 polynomial);
+u8 crc8(const u8 table[CRC8_TABLE_SIZE], u8 *pdata, size_t nbytes, u8 crc);
+
+#endif
diff --git a/preboot/crc8.c b/preboot/crc8.c
new file mode 100644
index 0000000..ffcab61
--- /dev/null
+++ b/preboot/crc8.c
@@ -0,0 +1,49 @@
+// SPDX-License-Identifier:     GPL-2.0+
+/*
+ * Copyright (C) 2017-2020 HabanaLabs Ltd.
+ * All Rights Reserved.
+ *
+ */
+
+#include <preboot.h>
+#include <crc8.h>
+
+/* This code was copied from from linux/lib/crc8.c */
+
+/**
+ * crc8_populate_msb - fill crc table for given polynomial in reverse bit order.
+ *
+ * @table:	table to be filled.
+ * @polynomial:	polynomial for which table is to be filled.
+ */
+void crc8_populate_msb(u8 table[CRC8_TABLE_SIZE], u8 polynomial)
+{
+	int i, j;
+	const u8 msbit = 0x80;
+	u8 t = msbit;
+
+	table[0] = 0;
+
+	for (i = 1; i < CRC8_TABLE_SIZE; i *= 2) {
+		t = (t << 1) ^ (t & msbit ? polynomial : 0);
+		for (j = 0; j < i; j++)
+			table[i+j] = table[j] ^ t;
+	}
+}
+
+/**
+ * crc8 - calculate a crc8 over the given input data.
+ *
+ * @table: crc table used for calculation.
+ * @pdata: pointer to data buffer.
+ * @nbytes: number of bytes in data buffer.
+ * @crc: previous returned crc8 value.
+ */
+u8 crc8(const u8 table[CRC8_TABLE_SIZE], u8 *pdata, size_t nbytes, u8 crc)
+{
+	/* loop over the buffer data */
+	while (nbytes-- > 0)
+		crc = table[(crc ^ *pdata++) & 0xff];
+
+	return crc;
+}
diff --git a/preboot/module_i2c_slave.c b/preboot/module_i2c_slave.c
index d7f6ca8..8eb3595 100644
--- a/preboot/module_i2c_slave.c
+++ b/preboot/module_i2c_slave.c
@@ -10,11 +10,15 @@
 #include <hl_i2c_slave_events.h>
 #include <hl_i2c_indirect.h>
 #include <boot_if.h>
+#include <crc8.h>
 
 #include "thermal.h"
 #include "module_i2c_slave.h"
 #include "i2cslave_indirect.h"
 #include "fit.h"
+
+u8 i2cslave_pec;
+static u8 pec_table[CRC8_TABLE_SIZE];
 
 static bool is_fit_checked;
 
@@ -81,8 +85,10 @@
 		ret = i2csi_ctrl_handle_status(cmd);
 		break;
 	default:
-		/* NA */
-		ret = 0;
+		if (HL_I2C_EVENT_IS_PEC(cmd))
+			ret = i2cslave_pec;
+		else
+			ret = 0;
 		break;
 	}
 
@@ -152,3 +158,13 @@
 	cpu_boot_sts_set(CPU_BOOT_STATUS_IN_WFE);
 	hl_dev_reset();	/* should not return */
 }
+
+void i2cslave_pec_init(void)
+{
+	crc8_populate_msb(pec_table, HL_I2C_CRC8_POLYNOMIAL);
+}
+
+void i2cslave_pec_digest(u8 val)
+{
+	i2cslave_pec = crc8(pec_table, &val, 1, i2cslave_pec);
+}
diff --git a/preboot/preboot.c b/preboot/preboot.c
index 2a478b7..85e1e1c 100644
--- a/preboot/preboot.c
+++ b/preboot/preboot.c
@@ -74,7 +74,7 @@
 		.check = module_i2c_s_check,
 		.disable = module_i2c_s_disable,
 		.run = module_i2c_s_run,
-		.enable = module_i2c_s_setup,
+		.enable = module_i2c_s_enable,
 		.run_fast = module_i2c_s_run_fast,
 		.teardown = module_i2c_s_teardown,
 		.run_noreturn = module_i2c_s_run_noret,
diff --git a/recovery/module_i2c_slave.c b/recovery/module_i2c_slave.c
index 8068154..9856a2f 100644
--- a/recovery/module_i2c_slave.c
+++ b/recovery/module_i2c_slave.c
@@ -10,6 +10,8 @@
 
 #include "module_i2c_slave.h"
 
+u8 i2cslave_pec;
+
 u8 i2cslave_handle_read_req(u8 cmd, u32 read_seq)
 {
 	u8 ret;
@@ -65,3 +67,7 @@
 }
 
 void i2cslave_reset_dev(void) {}
+
+void i2cslave_pec_init(void) {}
+
+void i2cslave_pec_digest(u8 val) {}
